lib/rosette/integrations/smartling-integration/smartling_completer.rb

```diff
@@ -1,49 +1,164 @@
 # encoding: UTF-8
 
+require 'thread'
+require 'concurrent'
+
 module Rosette
   module Integrations
     class SmartlingIntegration < Integration
       class SmartlingCompleter
 
-        attr_reader :configuration, :smartling_apis, :pullers
+        DEFAULT_THREAD_POOL_SIZE = 10
+
+        attr_reader :rosette_config, :repo_config
+        attr_reader :thread_pool_size
+
+        def initialize(rosette_config)
+          @rosette_config = rosette_config
+          @thread_pool_size = DEFAULT_THREAD_POOL_SIZE
+        end
 
-        def initialize(configuration, smartling_apis, pullers)
-          @smartling_apis = smartling_apis
-          @configuration = configuration
-          @pullers = pullers
+        def set_repo_config(repo_config)
+          @repo_config = repo_config
+          self
         end
 
-        def complete(repo_name, locales)
-          file_hash ||= Hash.new { |h, key| h[key] = [] }
-          file_uris_to_files = {}
+        def set_thread_pool_size(size)
+          @thread_pool_size = size
+          self
+        end
 
-          locales.each do |locale|
-            puller = puller_for_locale(locale)
-            completed_files = puller.completed_files_map[repo_name]
-            completed_files.each do |file|
-              file_uris_to_files[file.file_uri] ||= file
-              file_hash[file.file_uri] << locale
+        def complete
+          build_completion_maps.each do |file_uri, locales|
+            if locales.all? { |_, v| v.complete? }
+              begin
+                file = locales[repo_config.locales.first.code]
+                complete_log(file)
+                delete_file(file)
+              rescue => e
+                rosette_config.error_reporter.report_error(e)
+              end
             end
           end
+        end
 
-          file_hash.each do |file_uri, completed_locales|
-            if completed_locales.sort == locales.sort
-              Rosette.logger.info("Deleting file from Smartling: #{file_uri}")
-              file = file_uris_to_files[file_uri]
+        private
 
-              configuration.datastore.add_or_update_commit_log(
-                file.repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::TRANSLATED
-              )
+        def delete_file(file)
+          Retrier.retry(times: 3) do
+            smartling_api.delete(file.file_uri)
+          end.on_error(Exception).execute
+        end
 
-              smartling_apis[repo_name].delete(file_uri)
+        def complete_log(file)
+          rosette_config.datastore.add_or_update_commit_log(
+            file.repo_name, file.commit_id, nil, Rosette::DataStores::PhraseStatus::TRANSLATED
+          )
+        end
+
+        def build_completion_maps
+          if thread_pool_size > 1
+            build_completion_maps_asynchronously
```

Comment:
That's how we originally had it, but it made debugging tests more difficult. Trying to insert breakpoints and then having two threads (main and spawned) writing to the console was confusing. We could instead change this to `if thread_pool_size > 0` then build asynchronously, does that make more sense?